### ISSUE TITLE
Bug or intention? effective_get computes collection

### ISSUE THREAD

**Issue by mficek** [dask/dask#2247]
Dear dask authors, thanks for your effort and great job done on pushing dask where it is today!

I recently started using dask to load large CSV file, massage it a bit and store in hdf format. During this process I realized, that dask.dataframe.to_hdf function runs computation even if compute=False.
Please see the following code excerpt.

```
dask import dataframe as dd
from dask.diagnostics import ProgressBar, visualize
import dask
import pandas as pd
import numpy as np

d = pd.DataFrame({'a': np.random.randint(0,100, size=int(10e6))})
d.to_csv('tmp.csv')

pbar = ProgressBar()
pbar.register()

data = dd.read_csv('tmp.csv', blocksize=int(5e6))
save_data = data.to_hdf('tmp.hdf', '/data', compute=False)
dask.compute(save_data)
```
When you run it, you'll see that a progress bar is shown twice, once when `save_data = ... ` is executed and then when `dask.compute` is called. I traced the execution of `to_hdf` function and found that code, that triggers computation, is in `dask/utils.py`, namely the `effective_get` function:

```
def effective_get(get=None, collection=None):
    """Get the effective get method used in a given situation"""
    collection_get = collection._default_get if collection else None
    return get or _globals.get('get') or collection_get
```

Here, `effective_get` evaluates `collection` in `if` statement, but that translates to `collection.__len__` function and that one has to be evaluated, computed. Am I wrong, when I think, that there should be `if collection is not None`, or is there any intention behind this evaluation and we really want to know, if the collection is empty?

I'd propose that the `effective_get` looks like this:

```
def effective_get(get=None, collection=None):
    """Get the effective get method used in a given situation"""
    collection_get = collection._default_get if collection is not None else None
    return get or _globals.get('get') or collection_get
```

Such code works for me, because the CSV files of mine have tens of GBs and I store different slices of them in different HDF files. And reading such a big CSV file every time I want to "define" `data_save` variable without computation simply doesn't make any sense to me.

I'm happy to hear from you!

Python=2.7
pandas=0.19.2
dask=0.14.1
(but I don't think package versions play a significant role here)

**Comment by martindurant** [comment]
I think all the `to_*` functions compute immediately. 
These functions result in a concrete output rather than a further graph prescription, so they are different from all other dataframe operations. I suppose you are right, and in principle they could be performed asynchronously (I believe this option used to exist).

**Comment by mrocklin** [comment]
@mficek thank you for tracking this down and for writing up such a clear fail case.  At first glance I think that your understanding is correct and I think that your fix would be a good one.  Do you have any interest in submitting this as a PR?

Additionally we might also consider raising an exception on `dask.dataframe._Frame.__bool__` to catch these sorts of things earlier.

**Comment by mficek** [comment]
@mrocklin sure, I'll submit it as PR, no problem.
About raising an exception on `dask.dataframe._Frame.__bool__` I'm not so sure, have to check the code and think it through. Many times I use in pandas the expression `if dataframe: do something` and I mean it as it is written. The expression in `effective_get` looks like different case - simply to check, if argument to the function has been passed or not.

**Comment by jcrist** [comment]
Fixed by #2250.
